### PR TITLE
Refactor such that we can derive Encodable and Decodable for crypto crates APIs

### PR DESCRIPTION
**Pull request by joschisan**
Previously we derived encodable and decodable by hand for all the public types of tpe. However, with new crypto also built on bls12_381 like lightning-ng and ecash-ng that has more complex public structs this becomes increasingly cumbersome and repetitious. Hence, we refactor the code to be able to derive Encodable and Decodable for all future crypto crates built on bls12_381.

In a next step we want to do the same with serde Serialize and Deserialize by upstreaming an implementation to bls12_381 such that we can derive the traits as well. Notice that we duplicate the serde implementation for Scalar, G1, G2 as we copy it from tbs to core were we need it as well. We leave it in tbs though as the copy in core is supposed to be removed after we upstreamed it.

### CODE REVIEW

File: fedimint-core/src/encoding/threshold_crypto.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+    #[test_log::test]
+    fn test_ciphertext() {
+        let sks = threshold_crypto::SecretKeySet::random(1, &mut rand::thread_rng());
+        let pks = sks.public_keys();
+        let pk = pks.public_key();
+
+        let message = b"Hello world!";
+        let ciphertext = pk.encrypt(message);
+        let decryption_share = sks.secret_key_share(0).decrypt_share(&ciphertext).unwrap();
```
Comment (m1sterc001guy) [inline fedimint-core/src/encoding/threshold_crypto.rs:116]: Nit: use `expect` instead of `unwrap`

File: fedimint-core/src/encoding/bls12_381.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    ) -> Result<Self, crate::encoding::DecodeError> {
+        let mut bytes = [0u8; 48];
+
+        d.read_exact(&mut bytes)
+            .map_err(crate::encoding::DecodeError::from_err)?;
+
+        let point = G1Affine::from_compressed(&bytes);
+
+        if point.is_some().unwrap_u8() == 1 {
+            Ok(point.unwrap())
```
Comment (m1sterc001guy) [inline fedimint-core/src/encoding/bls12_381.rs:53]: Nit: use `expect`

File: fedimint-core/src/bls12_381_serde.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        let mut byte_array = [0u8; 32];
+
+        byte_array.copy_from_slice(&bytes);
+
+        let scalar = Scalar::from_bytes(&byte_array);
+        // FIXME: probably safe with public data, but doesn't look nice
+        if scalar.is_some().into() {
+            Ok(scalar.unwrap())
```
Comment (m1sterc001guy) [inline fedimint-core/src/bls12_381_serde.rs:36]: Nit: `expect`

**Review by m1sterc001guy — APPROVED** [review]
Someone with more context than me should also probably look, but LGTM

**Comment by dpc** [comment]
@elsirion

**Review by dpc — COMMENTED** [review]
On a lower level (mechanicaly what's goin on) it look OK, but on the higher level I'm lost.

The legacy vs new encoding seems like a complication, but for a good reason.

I also don't understand what's the status of "upstream serde support" or what exactly it means here.

@elsirion is probably best person to review

**Comment by elsirion** [comment]
Need to take a look, it's less of a priority than other things we might want to get into 0.3 since it's mostly useful for LN and e-cash NG, which are both longer-term research projects.

File: fedimint-core/src/encoding/bls12_381.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+impl Decodable for Scalar {
+    fn consensus_decode<D: std::io::Read>(
+        d: &mut D,
+        _modules: &ModuleDecoderRegistry,
+    ) -> Result<Self, crate::encoding::DecodeError> {
+        let mut bytes = [0u8; 32];
+
+        d.read_exact(&mut bytes)
```
Comment (elsirion) [inline fedimint-core/src/encoding/bls12_381.rs:19]: Why use a different function for decoding than encoding? You could use `<[u8; 32] as Decodable>::consensus_decode(…)` instead (or a variation thereof).

The problem here is that you make assumptions about the inner workings of `<[u8; 32] as Encodable>::consensus_encode` by re-implementing the corresponding decoding functionality manually.

File: fedimint-core/src/encoding/bls12_381.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            .map_err(crate::encoding::DecodeError::from_err)?;
+
+        let scalar = Scalar::from_bytes(&bytes);
+
+        if scalar.is_some().unwrap_u8() == 1 {
```
Comment (elsirion) [inline fedimint-core/src/encoding/bls12_381.rs:24]: Same improvement is possible as mentioned above.

File: fedimint-core/src/encoding/bls12_381.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    }
+}
+
+impl Encodable for G1Affine {
+    fn consensus_encode<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, std::io::Error> {
+        self.to_compressed().consensus_encode(writer)
+    }
+}
+
+impl Decodable for G1Affine {
```
Comment (elsirion) [inline fedimint-core/src/encoding/bls12_381.rs:40]: Same comments as for `Scalar` apply here too.

File: fedimint-core/src/encoding/bls12_381.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    }
+}
+
+impl Encodable for G2Affine {
+    fn consensus_encode<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, std::io::Error> {
+        self.to_compressed().consensus_encode(writer)
+    }
+}
+
+impl Decodable for G2Affine {
```
Comment (elsirion) [inline fedimint-core/src/encoding/bls12_381.rs:68]: Same comments as for `Scalar` apply here too.

File: fedimint-core/src/encoding/threshold_crypto.rs [hunk 8]
```diff
@@ -0,0 +1,121 @@
+use std::io::{Error, Read, Write};
+
+use threshold_crypto::group::Curve;
+use threshold_crypto::{G1Affine, G1Projective};
+
+use crate::encoding::{Decodable, DecodeError, Encodable};
+use crate::module::registry::ModuleDecoderRegistry;
+
+impl Encodable for threshold_crypto::PublicKeySet {
```
Comment (elsirion) [inline fedimint-core/src/encoding/threshold_crypto.rs:9]: Can't we just derive all this now? Care would need to be taken not to break anything, but worth looking into imo (follow-up, not this PR).

File: fedimint-core/src/encoding/threshold_crypto.rs [hunk 9]
```diff
@@ -0,0 +1,121 @@
+use std::io::{Error, Read, Write};
```
Comment (elsirion) [inline fedimint-core/src/encoding/threshold_crypto.rs:1]: I take it you just copied this whole file over.

**Review by dpc — DISMISSED** [review]

File: fedimint-core/src/bls12_381_serde.rs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        } else {
+            Err(D::Error::custom("Could not decode compressed G1Affine"))
+        }
+    }
+}
+
+macro_rules! impl_serde_g {
```
[the remark this answers is not in the document]
Comment (joschisan) [reply fedimint-core/src/bls12_381_serde.rs:38]: Done

File: fedimint-core/src/encoding/bls12_381.rs (hunk 4 above)
Comment (joschisan) [reply fedimint-core/src/encoding/bls12_381.rs]: Done

File: fedimint-core/src/encoding/bls12_381.rs (hunk 5 above)
Comment (joschisan) [reply fedimint-core/src/encoding/bls12_381.rs]: Done

File: fedimint-core/src/encoding/bls12_381.rs (hunk 6 above)
Comment (joschisan) [reply fedimint-core/src/encoding/bls12_381.rs:30]: Done

File: fedimint-core/src/encoding/bls12_381.rs (hunk 7 above)
Comment (joschisan) [reply fedimint-core/src/encoding/bls12_381.rs:48]: Done

File: fedimint-core/src/encoding/threshold_crypto.rs (hunk 9 above)
Comment (joschisan) [reply fedimint-core/src/encoding/threshold_crypto.rs:1]: Yes

**Review by dpc — APPROVED** [review]

File: fedimint-core/src/bls12_381_serde.rs [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+            s.serialize_bytes(&bytes)
+        }
+    }
+
+    pub fn deserialize<'d, D: Deserializer<'d>>(d: D) -> Result<Scalar, D::Error> {
+        let bytes: Vec<u8> = if d.is_human_readable() {
+            let deser: String = Deserialize::deserialize(d)?;
+            Vec::<u8>::from_hex(&deser).map_err(serde::de::Error::custom)?
+        } else {
+            Deserialize::deserialize(d)?
```
[the remark this answers is not in the document]
Comment (joschisan) [reply fedimint-core/src/bls12_381_serde.rs:22]: This is just copy and paste the old code from tbs for now.

[the remark this answers is not in the document]
Comment (elsirion) [reply fedimint-core/src/bls12_381_serde.rs:22]: I was confused, `serialize_bytes` always prepends the size, that's why deserializing a `Vec<u8>` works here.

**Review by elsirion — APPROVED** [review]

**Pull request merged**